juju/juju
# CAAS: Adds add-caas command, and AddCloud API.

## Description of change
This is all behind the "CAAS" feature flag.

Adds 'add-caas' command to add a cloud definition for a CAAS cluster.
We will reuse the Cloud type in state to store endpoints & creds for CAAS clusters.

- Adds "AddCloud" to Cloud facade. Revs Cloud version to 2 when CAAS feature flag is set
Traditional clouds are added via add-cloud before a controller is bootstrapped, and stored locally until bootstrap, at which point they're passed in via the bootstrap initialization data, and added directly (not using the API) by the controller during init. CAAS clouds are added to a pre-existing controller, so we needed to expose AddCloud in the Cloud facade.

What's missing, to be added in future branches:
- doesn't read CAAS config (e.g. $KUBECONFIG contents) to get creds or endpoint yet - plan is to add a clientconfig package to read that info
- no validation of given cloud type - plan is to add a CAASProvider to be the CAAS version of Environ
- needs an interactive version of the command (eventually)

## QA steps
with a bootstrapped controller: `juju add-caas anything somename`, then `juju clouds` should list it, and a cloud should show up in mongodb on the controller also, e.g.:

```
juju:PRIMARY> db.clouds.find()
{ "_id" : "localhost", "name" : "localhost", "type" : "lxd", "auth-types" : [ "interactive", "certificate" ], "endpoint" : ... etc ... , "txn-queue" : [ ] }
{ "_id" : "kubernetes", "name" : "kubernetes", "type" : "myk8s2", "auth-types" : [ "userpass" ], "endpoint" : "", "txn-revno" : NumberLong(2), "txn-queue" : [ "596eb0bb6792ac2655c9dbc3_9473df14" ] }
```

## Documentation changes

This will definitely affect user workflow, but is currently hidden behind a feature flag and the specific UX is up for discussion.
